Fix parsing record values containing colons

cyradotpink opened the pull request:
This PR is an attempt to fix #8257 and fix #10985 (which is duplicate-ish)

# Description
The parser currently doesn't know how to deal with colons appearing while lexing whitespace-terminated tokens specifying a record value. Most notably, this means you can't use datetime literals in record value position (and as a consequence, `| to nuon | from nuon` roundtrips can fail), but it also means that bare words containing colons cause a non-useful error message. 
![image](https://github.com/user-attachments/assets/f04a8417-ee18-44e7-90eb-a0ecef943a0f)

`parser::parse_record` calls `lex::lex` with the `:` colon character in the `special_tokens` argument. This allows colons to terminate record keys, but as a side effect, it also causes colons to terminate record *values*. I added a new function `lex::lex_n_tokens`, which allows the caller to drive the lexing process more explicitly, and used it in `parser::parse_record` to let colons terminate record keys while not giving them special treatment when appearing in record values.

This PR description previously said: *Another approach suggested in one of the issues was to support an additional datetime literal format that doesn't require colons. I like that that wouldn't require new `lex::lex_internal` behaviour, but an advantage of my approach is that it also newly allows for string record values given as bare words containing colons. I think this eliminates another possible source of confusion.* It was determined that this is undesirable, and in the current state of this PR, bare word record values with colons are rejected explicitly. The better error message is still a win.

# User-Facing Changes
In addition to the above, this PR also disables the use of "special" (non-item) tokens in record key and value position, and the use of a single bare `:` as a record key.

Examples of behaviour *before* this PR:
```nu
{ a: b } # Valid, same as { 'a': 'b' }
{ a: b:c } # Error: expected ':'
{ a: 2024-08-13T22:11:09 } # Error: expected ':'
{ :: 1 } # Valid, same as { ':': 1 }
{ ;: 1 } # Valid, same as { ';': 1 }
{ a: || } # Valid, same as { 'a': '||' }
```

Examples of behaviour *after* this PR:
```nu
{ a: b } # (Unchanged) Valid, same as { 'a': 'b' }
{ a: b:c } # Error: colon in bare word specifying record value
{ a: 2024-08-13T22:11:09 } # Valid, same as { a: (2024-08-13T22:11:09) }
{ :: 1 } # Error: colon in bare word specifying record key
{ ;: 1 } # Error: expected item in record key position
{ a: || } # Error: expected item in record value position
```

# Tests + Formatting
I added tests, but I'm not sure if they're sufficient and in the right place.

# After Submitting
I don't think documentation changes are needed for this, but please let me know if you disagree.

sholderbach commented:
Thanks for diving into this!

> but an advantage of my approach is that it also newly allows for string record values given as bare words containing colons. I think this eliminates another possible source of confusion.
> ![image](https://private-user-images.githubusercontent.com/12228688/350621596-12072839-55e2-4697-8185-657ce5645227.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE1NzM1NTEsIm5iZiI6MTcyMTU3MzI1MSwicGF0aCI6Ii8xMjIyODY4OC8zNTA2MjE1OTYtMTIwNzI4MzktNTVlMi00Njk3LTgxODUtNjU3Y2U1NjQ1MjI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIxVDE0NDczMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4Mzg3YzIyYjQ0MTE3MDk1YjRjZDA3MmZmMGVhZDE4NmZlZjViOGNmNTc5MDc5ODExM2NkYzJiZmI1ODI5NzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aWf84nTtZO5ZhgkIDpd-wIWSpVd6r0kYnwqeVJhOW3A)

I honestly think this could be a foot-gun if you mess up some whitespace or comma use. 
There are positions where bare-words should work very well like in the external command positions and in their arguments. In other positions I think accepting barewords is more a best effort thing, but the rules for when they are allowed should be relatively trivial to grok. Disallowing bare `:` in record key/value expressions seems reasonable to me. I think making our datetime literals work, seems fine as part of our grammar. (at the same time their fully qualified use with time may be somewhat rare)

cyradotpink commented:
That sounds reasonable to me! So would you be in favour of fixing the parsing issue as in this PR, but then explicitly disallow bare words in record values, giving a useful error message?

cyradotpink commented on crates/nu-parser/src/parser.rs:
```diff
@@ -5564,6 +5564,51 @@ pub fn parse_record(working_set: &mut StateWorkingSet, span: Span) -> Expression
+                    .find_position(|b| **b == b':')
+                    .map(|(i, _)| string_value.span.start + i);
+                if let Some(colon_position) = colon_position {
+                    Some(ParseError::InvalidLiteral(
```
I'm not sure this matches the intention of the `InvalidLiteral` error variant. Does another variant fit better or should we maybe introduce a new variant?

cyradotpink commented:
I'm still torn over whether I agree that colons in bare words in record value position should be an error. Constructs like `{ a:b:c }` look a little weird at first, but at least for me, already valid constructs like `{ a: b c: d }` were equally as unintuitive at first. The only footgunny behaviour I can really think of with this change is that some records stay valid but change meaning if you *forget* to put whitespace or a comma between key-value pairs, but that's a visually obvious mistake: `{ a: x b:y }` is very visually different from `{ a: xb:y }`. I don't think people need an error to catch that kind of mistake, somewhat analogously to how `[ a b ]` and `[ ab ]` can safely mean different things.

cyradotpink commented:
So - the way this works now is that I modified the lexing code a little bit to allow `parse_record` to kind of drive it more manually and switch between treating `:` specially and not. I also introduced a small *breaking change* by rejecting record keys that are just a single, bare colon (so `{ :: "x" }` no longer works; I actually just check if the key is a bare word containing any colons, it's just that this can only happen if it's a single bare colon) and rejecting record keys and values that are given as non-item-tokens (so `{ ;: "x"}`, `{ a: || }` etc. no longer work, I think personally I would consider it a bug that this worked before but that's a little subjective). I also explicitly detect and reject colons in bare word record values. I added some tests but I think they're kind of bad and maybe not in the right place?

sholderbach commented on crates/nu-parser/src/parser.rs:
```diff
@@ -5579,6 +5579,49 @@ pub fn parse_builtin_commands(
     }
 }
 
+fn check_record_key_or_value(
+    working_set: &StateWorkingSet,
+    expr: &Expression,
+    position: &str,
+) -> Option<ParseError> {
```
I am wondering if we have a more general version of this already to detect illegal chars that could be conditionally extended to include the colon for records (colons should also be disallowed for variable/parameter names due to its use in the type hints)

cyradotpink replied:
I could be missing something, but I don't think there is more existing code that I can use here. `check_record_key_or_value` produces an error if and only if the given Expression is a String or StringInterpolation *and* it is not a quoted string *and* contains a colon *in any of its string literal parts* (as in, it's important that interpolations like `{a: a('b:c')d}` remain allowed). This seems so inherently specific to me that I don't think there could exist much code I can reuse here. Looking at the examples you gave: colons are allowed in variable names (`let :a:b:: int = 10` just creates the variable named `$:a:b:`; the name and type are required to be whitespace-separated), and in signatures like `def [a:int:datetime:string] {}` the parser seems to just pick the name `$a` and the type `string`, ignoring valid types in between (which is.. suprising; we might want to revisit this some time?)

sholderbach commented on crates/nu-parser/tests/test_parser.rs:
```diff
@@ -2397,3 +2397,56 @@ mod operator {
+    #[rstest]
+    #[case(b"{ :: x }", "Invalid literal")] // Key is bare colon
+    #[case(b"{ a: x:y }", "Invalid literal")] // Value is bare word with colon
+    #[case(b"{ a: x('y'):z }", "Invalid literal")] // Value is bare string interpolation with colon
```
Just to clarify we don't have bare string interpolation, so that may need some clarification

cyradotpink replied:
I'm not sure what you mean here. By "bare string interpolation" I mean things like this: `{a: a(1 + 1)b}`, which is valid and produces the same record as `{a: 'a2b'}`. Am I not using the right terminology?

sholderbach replied:
huh, I thought that would not work in other contexts?

sholderbach commented on crates/nu-parser/src/lex.rs:
```diff
@@ -447,14 +447,59 @@ pub fn lex_signature(
+    pub span_offset: usize,
+}
+
+// Lex until the output is max_tokens longer than before the call, or until the input is exhausted.
+// The behaviour here is non-obvious (maybe non-useful) iff your additional_whitespace doesn't include newline:
+// If you pass `output` in a state where the last token is an Eol, this might *remove* tokens.
```
I am honestly not familiar enough with the lexing step we have to make an substantive comment to this, but I would suggest making the key parts here a doccomment for folks new to the lexer (like me :D)

cyradotpink replied:
Agreed, good call :)

sholderbach reviewed (commented):
Thanks for iterating on this, I am not really versed in our lexing and the follow on interactions during parsing, so I can't really give you an expert look at your changes.

For the general public it may be worth compiling the changes to allowed/disallowed records for the release notes (and any other reviewers jumping in)

cyradotpink commented:
Finally picking this back up, sorry for the delay. I updated the description with relevant information about the current state of this PR.

cyradotpink commented:
> For the general public it may be worth compiling the changes to allowed/disallowed records for the release notes (and any other reviewers jumping in)

Good idea - I added some examples of what changed to the "User-Facing Changes" section in the initial comment.

sholderbach commented:
OK, to update here, I am fully suppportive of the top level syntax changes you arrived at.
I would defer to the parser nerds if there may be some pathological cases we could run into with the now stateful lexing for the `lex_n_tokens` thingie. (be it bugs or some gnarly worst case performance)

Otherwise we should probably go ahead and merge and see the impact in the wild and make sure to replicate the strictness in the `new-nu-parser` reimplementation

sholderbach commented:
Let's give it a shot!

KAAtheWiseGit commented:
Thanks a lot for this PR!
